### PR TITLE
Prepare publishing of blinded balance proofs

### PR DESCRIPTION
**Pull request by konradkonrad**
This is progress towards #3273

The PR adds the necessary pieces to publish blinded balance proofs to the global matrix room `monitoring`. 

In particular:
- add a new event `EventNewBalanceProofReceived`
- add a state diffing mechanism, that detects new balance proofs from the global `StateManager.dispatch` method, and injects the new events
- add `transport.send_global` calls in the `RaidenMonitoringEventHandler`
- add a matrix integration test for the global messages to be sent
- add `EventNewBalanceProofReceived` to `test_mediated_transfer_events` happy case

It also changes the matrix transport behavior for `send_global`: previously `send_global` would raise `AssertionError`, if a room target was not set up beforehand. Now the target room will be joined automatically. This is meant to avoid the overhead of checking for the room existence in sending code, or catching the Assertion. If this last change is contentious, we could add a non-asserting `is_global_room_available(room_name)` method instead.

### CODE REVIEW

**Comment by konradkonrad** [comment]
This is still missing recovery at startup!

**Comment by konradkonrad** [comment]
One more comment: ideally publishing to the monitoring service would block further processing of the channel until the message was sent(received). The idea behind that is, that monitoring is needed for safety, and therefore the channel's protocol should **not** proceed until monitoring was successfully requested (note: current MS design does not give any feedback on a request).

**Comment by ulope** [comment]
The question is what form of feedback would make sense since there should ideally be many monitoring services watching the broadcast channel.

**Comment by konradkonrad** [comment]
> The question is what form of feedback would make sense since there should ideally be many monitoring services watching the broadcast channel.

This PR is probably the wrong place to discuss this. 

Nevertheless: In a long term view, I could see MS's publishing some batched "acknowledged balance_hashes" to the same room every once in a while (tbd.). Since MS's will compete in QoS, clients would whitelist some MS addresses and wait for N ack's before proceeding.

File: raiden/raiden_event_handler.py [hunk 1]
```diff
@@ -546,18 +547,28 @@ class RaidenMonitoringEventHandler(RaidenEventHandler):
...
 
-    def handle_received_payment(
+    def handle_received_balance_proof(
             self,
             raiden: RaidenService,
-            payment_received: EventPaymentReceivedSuccess,
+            new_balance_proof_event: EventNewBalanceProofReceived,
     ):
         log.info(
-            'Received payment, sending balance proof to Monitoring Service',
```
Comment (hackaugusto) [inline raiden/raiden_event_handler.py]: receiving a payment and a new balance proof are not the same thing, please don't replace one with other

File: raiden/transfer/architecture.py [hunk 2]
```diff
@@ -264,6 +280,14 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         assert isinstance(iteration, TransitionResult)
 
+        # if any balance proofs have changed, create events
+        balance_proofs = self.detect_balance_proof_change(
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: I don't think this needs to be part of the state machine, this code is not important for the protocol or restartability, so it can safely be called from somewhere else

File: raiden/transfer/architecture.py [hunk 3]
```diff
@@ -272,6 +296,50 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         return events
 
+    def detect_balance_proof_change(self, old_state, current_state):
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: this is quite hard to follow, could you please simplify this code?

File: raiden/transfer/architecture.py (hunk 2 above)
Comment (konradkonrad) [reply raiden/transfer/architecture.py]: where would you rather see this integrated?

File: raiden/raiden_event_handler.py (hunk 1 above)
Comment (konradkonrad) [reply raiden/raiden_event_handler.py]: as you can probably see, this is the old version, and I changed it to balance proofs

File: raiden/transfer/architecture.py (hunk 3 above)
Comment (konradkonrad) [reply raiden/transfer/architecture.py]: I tried to make it more explicit!

File: raiden/raiden_event_handler.py (hunk 1 above)
Comment (konradkonrad) [reply raiden/raiden_event_handler.py]: another note for context: the above lines are from `RaidenMonitoringEventHandler` which was previously only a stub, (incorrectly) assuming to operate on `EventPaymentReceivedSuccess`.

File: raiden/transfer/architecture.py [hunk 4]
```diff
@@ -272,6 +295,66 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
...
+        """Compare two states for any changed received balance_proofs"""
+        if old_state == current_state:
+            return
+        # Assuming, there is no transition None->new balance proof:
+        if not hasattr(old_state, 'identifiers_to_paymentnetworks'):
+            return
+        for old_payment_network_identifier, current_payment_network_identifier in zip(
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: I don't think this works as you expect, `identifiers_to_paymentnetworks` are dictionaries, and iterating over them returns the dict keys in order of insertion (python 3.6+):

```python
>>> d = {1:1, 2:2}
>>> d2 = {2:2, 1:1}
>>> list(zip(d, d2))
[(1, 2), (2, 1)]
```

Which for all purposes is undefined at this point.

File: raiden/transfer/architecture.py [hunk 5]
```diff
@@ -272,6 +295,66 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         return events
 
+    def detect_balance_proof_change(self, old_state, current_state):
+        """Compare two states for any changed received balance_proofs"""
+        if old_state == current_state:
+            return
+        # Assuming, there is no transition None->new balance proof:
+        if not hasattr(old_state, 'identifiers_to_paymentnetworks'):
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: Checking if the attribute exists is not sufficient, you should check if it's not `None`, `getattr(.., .., None)` is probably best here.

File: raiden/transfer/architecture.py (hunk 4 above)
Comment (hackaugusto) [reply raiden/transfer/architecture.py]: Since this is getting the *new* balance proofs, it just needs to get the keys from `current_state.identifiers_to_paymentnetworks`, if there is no corresponding key in `old_state.identifiers_to_paymentnetworks` then we know all BPs are new.

File: raiden/transfer/architecture.py [hunk 6]
```diff
@@ -272,6 +295,66 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         return events
 
+    def detect_balance_proof_change(self, old_state, current_state):
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: please, move this to the views module and add types

File: raiden/transfer/architecture.py [hunk 7]
```diff
@@ -264,6 +280,13 @@ def dispatch(self, state_change: StateChange) -> List[Event]:
 
         assert isinstance(iteration, TransitionResult)
 
+        # if any balance proofs have changed, create events
+        for balance_proof in self.detect_balance_proof_change(
+                self.current_state,
+                iteration.new_state,
+        ):
+            iteration.events.append(EventNewBalanceProofReceived(balance_proof))
```
Comment (hackaugusto) [inline raiden/transfer/architecture.py]: This is no place for this logic, the state manager does all it needs to do already. You can just call `detect_balance_proof_change` after a state change is processed.

File: raiden/transfer/architecture.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/transfer/architecture.py]: sorry, I missed this question: https://github.com/raiden-network/raiden/pull/3452#discussion_r255958278

File: raiden/transfer/mediated_transfer/events.py [hunk 8]
```diff
@@ -562,6 +562,22 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
         return restored
 
 
+class EventNewBalanceProofReceived(Event):
+    """ Event for newly received balance proofs. Useful for notifying monitoring services. """
+
+    def __init__(self, balance_proof):
+        self.balance_proof = balance_proof
+
+    def to_dict(self):
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: Please add return type.

File: raiden/transfer/mediated_transfer/events.py [hunk 9]
```diff
@@ -562,6 +562,22 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
...
+        return {
+            'balance_proof': self.balance_proof.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, data):
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: `data` param type and return type.

Comment (konradkonrad) [reply raiden/transfer/mediated_transfer/events.py]: done

File: raiden/transfer/mediated_transfer/events.py (hunk 8 above)
Comment (konradkonrad) [reply raiden/transfer/mediated_transfer/events.py]: done

File: raiden/utils/typing.py [hunk 10]
```diff
@@ -131,6 +131,12 @@
     bound='raiden.messages.RequestMonitoring',
 )
 
+
+T_EventNewBalanceProofReceived = TypeVar(
+    'T_EventNewBalanceProofReceived',
+    bound='raiden.transfer.mediated_transfer.events.EventNewBalanceProofReceived',
+)
```
Comment (rakanalh) [inline raiden/utils/typing.py]: Please remove

File: raiden/transfer/mediated_transfer/events.py [hunk 11]
```diff
@@ -562,6 +564,25 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
...
+    def to_dict(self) -> Dict:
+        return {
+            'balance_proof': self.balance_proof.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(
+            cls: Type[T_EventNewBalanceProofReceived],
+            data: Dict,
+    ) -> T_EventNewBalanceProofReceived:
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: I am not sure why you had to declare a type for this. The type would be `EventNewBalanceProofReceived`, no need to declare yet another one in typing.

File: raiden/transfer/mediated_transfer/events.py [hunk 12]
```diff
@@ -562,6 +564,25 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
         return restored
 
 
+class EventNewBalanceProofReceived(Event):
+    """ Event for newly received balance proofs. Useful for notifying monitoring services. """
+
+    def __init__(self, balance_proof: BalanceProofSignedState) -> None:
+        self.balance_proof = balance_proof
+
+    def to_dict(self) -> Dict:
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: `Dict[str, Any]`

File: raiden/transfer/mediated_transfer/events.py [hunk 13]
```diff
@@ -562,6 +564,25 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
         return restored
 
 
+class EventNewBalanceProofReceived(Event):
+    """ Event for newly received balance proofs. Useful for notifying monitoring services. """
+
+    def __init__(self, balance_proof: BalanceProofSignedState) -> None:
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: No need for `-> None`

File: raiden/transfer/mediated_transfer/events.py [hunk 14]
```diff
@@ -562,6 +564,25 @@ def from_dict(cls, data: Dict[str, Any]) -> 'EventUnlockFailed':
...
+        self.balance_proof = balance_proof
+
+    def to_dict(self) -> Dict:
+        return {
+            'balance_proof': self.balance_proof.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(
+            cls: Type[T_EventNewBalanceProofReceived],
```
Comment (rakanalh) [inline raiden/transfer/mediated_transfer/events.py]: We don't give a type for `cls` and `self`.

File: raiden/transfer/mediated_transfer/events.py (hunk 11 above)
Comment (konradkonrad) [reply raiden/transfer/mediated_transfer/events.py]: please follow the conversation at https://github.com/raiden-network/raiden/pull/3374#discussion_r253424031 and https://github.com/raiden-network/raiden/pull/3374/commits/69831a8fab12ea38c4ab98ae1122792d7a5d28ae
tldr; class is not available inside class definition

File: raiden/utils/typing.py (hunk 10 above)
Comment (konradkonrad) [reply raiden/utils/typing.py]: please follow the conversation at https://github.com/raiden-network/raiden/pull/3374#discussion_r253424031 and https://github.com/raiden-network/raiden/pull/3374/commits/69831a8fab12ea38c4ab98ae1122792d7a5d28ae

File: raiden/transfer/mediated_transfer/events.py (hunk 13 above)
Comment (konradkonrad) [reply raiden/transfer/mediated_transfer/events.py]: https://github.com/raiden-network/raiden/pull/3374#discussion_r253011733

File: raiden/transfer/mediated_transfer/events.py (hunk 11 above)
Comment (rakanalh) [reply raiden/transfer/mediated_transfer/events.py]: None of our other events / state changes / states declare this. Otherwise we would have hundreds of types. Please follow the same convention used in other events in the same file.
